[FEATURE] 관리자의 Q&A 관리 기능 구현

JakePark929 opened the pull request:
## 📝 PR Summary

관리자의 Q&A 관리 기능을 구현합니다.
관리자가 답변을 하면 자동으로 답변 완료 상태가 됩니다.

- `CoreApiApplication` 에 `CommandLineRunner` 를 사용하여 local profile 일 때, 기본 유저와 Q&A 데이터를 추가하는 기능을 추가했습니다.
- `Qna` 도메인에 불필요한 setter 를 제거하고, update 메소드를 추가했습니다.
- 관리자 용 `Q&A 전체 조회`, `내 Q&A 조회`, `Q&A 업데이트` 기능이 추가되었습니다.
  > Q&A 전체 조회와 내 Q&A 조회는 커버링 인덱스 방식으로 페이징 처리 됩니다.
- 관리자 Q&A 관리와 관련한 비즈니스 로직 단위의 테스트 코드가 추가되었습니다.
  > 테스트 코드 추가에 따라 properties 에 test profile 이 추가되었습니다.

#### 🌲 Working Branch

feat/#30-admin-qna

#### 🌲 TODOs

- [x] 질문 전체 조회
- [x] 내 답변 리스트 조회
- [x] 내 답변 등록
- [x] 진행 상태 수정

### Related Issues

resolve #30 

통합 테스트 추가하고, 비즈니스 레이어에서 테스트 코드도 구현해 봤는데 이거 하느라 시간이 많이 들었네요..
테스트 코드 경험이 많지 않아서 아직 부족합니다. 많은 의견 부탁 드립니다 :)

JakePark929 committed:
[#30] feat : 관리자의 Q&A 관리 기능 구현

- `CoreApiApplication` 에 `CommandLineRunner` 를 사용하여 local profile 일 때, 기본 유저와 Q&A 데이터를 추가하는 기능을 구현했습니다.
- `Qna` 도메인에 불필요한 setter 를 제거하고, update 메소드를 추가했습니다.
- 관리자용 `Q&A 전체조회`, `내 Q&A 조회`, `Q&A 업데이트` 기능이 추가되었습니다.
  > `Q&A 전체조회`와 `내 Q&A 조회`는 커버링인덱스 방식으로 페이징처리 됩니다.
- 관리자 Q&A 관리와 관련한 비즈니스 로직단위의 테스트 코드가 추가되었습니다.
  > 테스트 코드 추가에 따라 properties 에 test profile 이 추가되었습니다.

JakePark929 committed:
[#30] fix : 초기 유저 데이터 변경

- local 초기 DB sub id 와 email 을 숨김처리 했습니다. 본인 이메일로 사용가능합니다.

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/CoreApiApplication.java:
```diff
@@ -1,15 +1,98 @@
 @ConfigurationPropertiesScan
 @SpringBootApplication
 public class CoreApiApplication {
+    @Value("${oauth2.admin.sub-id}")
+    String adminSubId;
+    @Value("${oauth2.admin.email}")
+    String adminEmail;
+    @Value("${oauth2.user.sub-id}")
+    String userSubId;
+    @Value("${oauth2.user.email}")
+    String userEmail;
```
접근제어자가 필요하지 않을까요?

donsonioc2010 replied:
이런 방식도 좋고, Properties라고해서, 클래스를 만들고 주입하는 방식도 괜찮을 것 같습니다 :)

JakePark929 replied:
properties 만들어 보았습니다! 좋은 정보 감사합니다 👍

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/CoreApiApplication.java:
```diff
@@ -1,15 +1,98 @@
+                    .subId(userSubId)
+                    .email(userEmail)
+                    .name("사용자")
+                    .gender(UserGender.MAN)
+                    .thumbnail("https://lh3.googleusercontent.com/")
+                    .role(UserRole.USER)
+                    .status(UserStatus.ACTIVE)
+                    .build();
+
+            userRepository.save(user1);
+
+            Qna qna1 = Qna.builder()
+                    .qnaUser(user1)
+                    .title("문의문의 나문의")
+                    .content("호박고구마")
+                    .progressStatus(QnaProgressStatus.UNPROCESSED)
+                    .answerStatus(AnswerStatus.NO_REPLY)
+                    .build();
+
+            Qna qna2 = Qna.builder()
+                    .qnaUser(user1)
+                    .title("문의문의 나문의2")
+                    .content("호박고구마!!")
+                    .progressStatus(QnaProgressStatus.UNPROCESSED)
+                    .answerStatus(AnswerStatus.NO_REPLY)
+                    .build();
+
+            Qna qna3 = Qna.builder()
+                    .qnaUser(user1)
+                    .title("문의문의 나문의3")
+                    .content("호박고구마!!!")
+                    .progressStatus(QnaProgressStatus.UNPROCESSED)
+                    .answerStatus(AnswerStatus.NO_REPLY)
+                    .build();
+
+            qnaRepository.save(qna1);
+            qnaRepository.save(qna2);
+            qnaRepository.save(qna3);
+        };
+    }
```
Core-Api 루트어플리케이션에 있는거보다 분리하는게 좋을것 같습니다 ㅋㅋ

JakePark929 replied:
이 부분도 component 로 분리 하였습니다.
db에 추가되는 출처 확인 때문에 메인에 명시적으로 써준다는 얘기를 얼핏 들었는데
얼핏 들은거라 확실하지 않습니다 🤣

컴포넌트로 분리하니 깔끔해졌네요!

donsonioc2010 commented on build.gradle:
```diff
@@ -33,6 +33,9 @@ subprojects {
         testCompileOnly 'org.projectlombok:lombok'
         testAnnotationProcessor 'org.projectlombok:lombok'
+
+        testImplementation 'org.junit.jupiter:junit-jupiter-api'
+        testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine'
```
혹시 Root의 Build.gradle에 추가한 이유가 있을까요?

제가 알기로는 spring-boot-starter-test내부에 이미 junit이 포함이 되어있는걸로 알다보니, 불필요한 작업이 아닌가 생각이 들어서 그렇습니다

JakePark929 replied:
테스트 실행이 안되서 여기저기 설정하다 보니 그대로 두고 진행했었네요! 이 부분은 원복해 놓았습니다!

donsonioc2010 commented on core/core-api/build.gradle:
```diff
@@ -8,6 +8,7 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-web'
     implementation 'org.springframework.boot:spring-boot-starter-validation'
     implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
+    testImplementation 'org.springframework.boot:spring-boot-starter-test'
```
이미 Root.gradle에 추가가 되어있는 의존성인데, 추가한 이유가 있을까요?

JakePark929 replied:
이 부분도 위 사유와 동일합니다~ :)

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/qna/business/QnaAdminBusiness.java:
```diff
@@ -0,0 +1,69 @@
+        return qnaService.updateQnaFromAdmin(userDto.getId(), qnaId, request);
+    }
+
+    private QnaProgressStatus validateProgressStatus(String progressStatus) {
+        if (ObjectUtils.isEmpty(progressStatus)) {
+            return null;
+        }
+
+        return QnaProgressStatus.valueOf(progressStatus);
+    }
```
해당 메소드는 QnaProgressStatus에서 가지고 있는게 좋아보입니다.

또한, Null을 반환하기 보단, 문자열을 잘못 제공한 것이기에 Exception을 주거나, 혹은 Default상태값을 주는게 좋아 보일것 같은 생각이 드네요.

JakePark929 replied:
처음부터 enum 으로 받으니 이 부분은 삭제 되었습니다.
할 때는 뭔가 안 되서 추가했던 것 같은데 이제 잘 되네요! 🤣🤣

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/qna/business/QnaAdminBusiness.java:
```diff
@@ -0,0 +1,69 @@
+     * @param pageable 페이지 객체
+     * @return Q&A 리스트 with page
+     */
+    public Page<QnaResponse> getQnas(String progressStatus, Pageable pageable) {
+
+        return qnaService.getAllQna(null, validateProgressStatus(progressStatus), pageable);
+    }
```
해당 ProgressStatus가 문자열로 오다 보니 하위에 private으로 메소드를 만드셨다 판단이 드네요,

차라리 Enum타입으로 처음부터 받으셨으면 하는 생각이 드네용!

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/qna/controller/QnaAdminController.java:
```diff
@@ -0,0 +1,59 @@
+    @GetMapping
+    public ApiResponse<Page<QnaResponse>> getQnas(
+            @UserSession UserDto userDto,
+            @RequestParam(required = false) String progressStatus,
```
RequestParam이 안들어 올 수 있는 케이스가 존재하는건가요?

JakePark929 replied:
status가 안 들어오면 모든 상태를 검색하도록 하는 경우를 생각했는데요~ 
required 를 안 하면 필수로 받아야 되서 전체 조회를 못했던 것 같습니다.
다른 좋은 방식 있을까요? :)

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/qna/service/QnaService.java:
```diff
@@ -55,4 +111,51 @@ public Qna insertQna(UserDto userDto, QnaAddRequest request) {
+
+    // 답변 업데이트 메소드
+    private Qna update(User adminUser, Qna qnaEntity, QnaUpdateRequest request) {
+        Class<? extends QnaUpdateRequest> requestClass = request.getClass();
+        Field[] fields = requestClass.getDeclaredFields();
```
이렇게 사용해야 하는 이유가 있으실까요?

Setter의 제한이 필요하면, 필요한 부분의 책임이 있는 메소드를 만드는게 좋지 않을까 하는 생각이 드네용 :)

JakePark929 replied:
그렇네요! 도메인에서 메소드 만들어 봤습니다! 훨씬 좋네요 😄

donsonioc2010 commented on core/core-api/src/main/resources/application.yml:
```diff
@@ -24,6 +38,12 @@ oauth2:
+    email: ${OAUTH2_ADMIN_EMAIL}
+  user:
+    sub-id: ${OAUTH2_USER_SUB_ID}
+    email: ${OAUTH2_USER_EMAIL}
```
사실상 동일한 역할입니당 :)
```suggestion
spring.application.name: core-api
spring.profiles.active: local

spring:
  config:
    import:
      - optional:file:.env[.properties]
      - db-core.yml
  mvc.throw-exception-if-no-handler-found: true
#  web.resources.add-mappings: false

server:
  tomcat:
    max-connections: 20000
    threads:
      max: 600
      min-spare: 100

oauth2:
  google:
    client-id: ${OAUTH2_GOOGLE_CLIENT_ID}
    client-secret: ${OAUTH2_GOOGLE_CLIENT_SECRET}
    redirect-uri: ${OAUTH2_GOOGLE_REDIRECT_URI}
    authorization-uri: https://accounts.google.com/o/oauth2/auth
    token-uri: https://oauth2.googleapis.com/token
    user-info-uri: https://www.googleapis.com/oauth2/v3/userinfo

  admin:
    sub-id: ${OAUTH2_ADMIN_SUB_ID}
    email: ${OAUTH2_ADMIN_EMAIL}
  user:
    sub-id: ${OAUTH2_USER_SUB_ID}
    email: ${OAUTH2_USER_EMAIL}
```

JakePark929 replied:
이 부분도 마찬가지로 환경 변수 때문에 테스트 진행이 안 되서 설정했는데 아래 알려주신 정보로 분리하니 정상 동작합니다! 덕분에 다시 원복 해도 되겠네요~

donsonioc2010 commented on core/core-api/src/main/resources/application.yml:
```diff
@@ -52,3 +72,34 @@ spring.config.activate.on-profile: staging
 ---
 spring.config.activate.on-profile: live
 
+---
+spring:
+  config.activate.on-profile: test
+
+oauth2:
+  google:
+    client-id: "client-id"
+    client-secret: "client-secret"
+    redirect-uri: "redirect-uri"
+    authorization-uri: https://accounts.google.com/o/oauth2/auth
+    token-uri: https://oauth2.googleapis.com/token
+    user-info-uri: https://www.googleapis.com/oauth2/v3/userinfo
+
+token:
+  secret:
+    key: "test token"
+  access-token:
+    plus-hour: 1
+  refresh-token:
+    plus-hour: 12
+
+storage:
+  datasource:
+    core:
+      driver-class-name: org.h2.Driver
+      jdbc-url: jdbc:h2:mem:core;MODE=MySQL;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE
+      username: sa
+      pool-name: core-db-pool
+      data-source-properties:
+        rewriteBatchedStatements: true
+
```
Test케이스에서만 실행되게 하고싶은 경우 test디렉토리 하위에 resource에서 application.yml을 제작하는 경우 해당 application.yml이 우선적용됩니다. 같이 참고하시면 좋을 것 같습니다. 👍

JakePark929 replied:
아주 도움되는 정보네요! 감사합니다 👍

donsonioc2010 reviewed (commented):
설정관련해서 다시 바꾸셔야 할게 있다고 생각되서 일단 상태는 코멘트로 두겠습니당 :)

설정 바꿀부분들이 있으면, 언제든지 물어보셔도 괜찮습니다. 저도 모르는거면 같이 찾고 알고있는거면 여유될때 바로 답장드릴 수 있도록 할게요  ㅎㅎ :)

언제나 고생많으십니당!

birdieHyun commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/qna/dto/request/QnaUpdateRequest.java:
```diff
@@ -0,0 +1,17 @@
+import lombok.Getter;
+
+@Getter
+@AllArgsConstructor
```
정적 팩터리 메서드로 객체를 생성하는 것으로 보이는데, 이럴 경우 생성자는 막아주는건 어떨까요?

JakePark929 replied:
이 부분 적용해봤습니다.
자주 활용하겠습니다 😄

JakePark929 committed:
[#30] refactor : 테스트 설정 원상복구, data loader 컴포넌트 추가 외 리뷰사항 반영

- 테스트를 위해 추가했던 설정들은 불필요하여 원복하고, 테스트용 yml 파일이 추가되었습니다.
- 메인에 있던 data loader 코드를 컴포넌트화 하고 properties 클래스를 추가했습니다.
- 불필요하게 길었던 Q&A 업데이트 코드를 간소화 했습니다.

JakePark929 committed:
Merge remote-tracking branch 'origin/dev' into feat/#30-admin-qna

donsonioc2010 reviewed (approved):
LGTM~
